### PR TITLE
Login API route

### PR DESCRIPTION
**Pull request by pontepotchet**
Me and @merrilytan discussed using api routes with naming conventions like /users/new and /users/:userid for registration and login, but since we are only using email addresses to sign in, I couldn't think of what userid we could supply to this route, other than the email address itself. 
I then found examples online of people using routes like /api/login and /api/register, so I stuck with these for now. Can someone advise me?

### CODE REVIEW

**Comment by aeciorc** [comment]
> Me and @merrilytan discussed using api routes with naming conventions like /users/new and /users/:userid for registration and login, but since we are only using email addresses to sign in, I couldn't think of what userid we could supply to this route, other than the email address itself.

I think POST:/users/:userid for login wouldn't make sense since the id is something the server knows about, but not the client. GET:/users/:userid will be used later for fetching the user profile, the id being the mongodb _id.  
> I then found examples online of people using routes like /api/login and /api/register, so I stuck with these for now. Can someone advise me?

How about POST:*/users* for registration and POST:*users/login*  ?

**Review by aeciorc — COMMENTED** [review]
The PR looks great. I would avoid using the error code 500 since that's usually for an actual server error.  And for keys in config/keys.js, I think you should put the actual value in the .env file and get them using process.env

File: server/config/keys.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
 module.exports = {
-	mongoURI: "mongodb+srv://dbAdmin:teamcactus@hatch-ptlkp.mongodb.net/test"
+	mongoURI: "mongodb+srv://dbAdmin:teamcactus@hatch-ptlkp.mongodb.net/test",
```
Comment (skassam21) [inline server/config/keys.js]: For local development, you should use a local server no? 
We'd probably want to make these environment variables (in .env) and then gitignore those files

File: server/config/passport.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+const keys = require("../config/keys");
+
+const opts = {};
+opts.jwtFromRequest = ExtractJwt.fromAuthHeaderAsBearerToken();
+opts.secretOrKey = keys.secretOrKey;
+
+module.exports = passport => {
+	passport.use(
+		new JwtStrategy(opts, (jwt_payload, done) => {
+			User.findById(jwt_payload.id)
```
Comment (skassam21) [inline server/config/passport.js:16]: I believe you are actually not saving your id in the JWT payload, but the email? Either choose to save id (which is standard and more efficient), or you need to fetch user by email here

File: server/routes/register.js [hunk 3]
```diff
@@ -30,12 +28,13 @@ router.post('/register', async (req, res) => {
             await user.save();
             res.status(201).send(user);
```
Comment (skassam21) [inline server/routes/register.js:29]: I usually send a JWT token in register as well, so the user can be logged in after register as well

**Review by skassam21 — COMMENTED** [review]
Overall looks good, there may be a bug in the passport middleware, probably something you couldn't find or test till you added a protected route

**Comment by pontepotchet** [comment]
> I would avoid using the error code 500 since that's usually for an actual server error.

@aeciorc What kind of error would it be if the server encountered some problem while using a password hashing utility or saving an entry to the database?

**Comment by aeciorc** [comment]
> @aeciorc What kind of error would it be if the server encountered some problem while using a password hashing utility or saving an entry to the database?

For hashing, it shouldn't really happen unless it's a bad input, you could use 422 for that. For a database error, I think 503 would be appropriate since it's usually a temporary issue.
It's not really something you should worry too much about, but it's better not to return a 500 on purpose.

**Pull request merged**